nodebuilder/store.go

```diff
@@ -182,3 +182,61 @@ func indexPath(base string) string {
 func dataPath(base string) string {
 	return filepath.Join(base, "data")
 }
+
+// constraintBadgerConfig returns BadgerDB configuration optimized for low memory usage and more frequent
+// compaction which prevents memory spikes.
+// This is particularly important for LNs with restricted memory resources.
+//
+// With the following configuration, a LN uses up to 300iB of RAM during initial sync/sampling
+// and up to 200MiB during normal operation. (on 4 core CPU, 8GiB RAM droplet)
+//
+// With the following configuration and "-tags=jemalloc", a LN uses no more than 180MiB during initial
+// sync/sampling and up to 100MiB during normal operation. (same hardware spec)
+// NOTE: To enable jemalloc, build celestia-node with "-tags=jemalloc" flag, which configures Badger to
+// use jemalloc instead of Go's default allocator.
+//
+// TODO(@Wondertan): Consider alternative less constraint configuration for FN/BN
+// TODO(@Wondertan): Consider dynamic memory allocation based on available RAM
+func constraintBadgerConfig() *dsbadger.Options {
+	opts := dsbadger.DefaultOptions // this must be copied
+	// ValueLog:
+	// 2mib default => 500b - makes sure headers and samples are stored in value log
+	// This *tremendously* reduces the amount of memory used by the node, up to 10 times less during
+	// compaction
+	opts.ValueThreshold = 500
+	// make sure we don't have any limits for stored headers
+	opts.ValueLogMaxEntries = 100000000
+	// run value log GC more often to spread the work over time
+	opts.GcInterval = time.Minute * 1
+	// default 0.5 => 0.125 - makes sure value log GC is more aggressive on reclaiming disk space
+	opts.GcDiscardRatio = 0.125
+	// Snappy saves us ~200MiB of disk space on the mainnet chain to the commit's date
+	// TODO(@Wondertan): Does it worth the overhead?
+	opts.Compression = options.Snappy
+	opts.BlockCacheSize = 16 << 20
+
+	// MemTables:
+	// default 64mib => 16mib - decreases memory usage and makes compaction more often
+	opts.MemTableSize = 16 << 20
+	// default 5 => 3
+	opts.NumMemtables = 3
+	// default 5 => 3
+	opts.NumLevelZeroTables = 3
+	// default 15 => 5 - this prevents memory growth on CPU constraint systems by blocking all writers
+	opts.NumLevelZeroTablesStall = 5
+
+	// Compaction:
+	// Dynamic compactor allocation
+	compactors := runtime.NumCPU() / 2
```

Comment:
There will be same amount of work requited for compaction (aka cpu time). So if we limit compactors to use half of the CPU cores, it will be just slower. Also would add risk of hangups if it happen to be too slow.

Reply:
Limiting the number of compactors prevents us from running all compactors simultaneously and spiking RAM usage. It's okay that compaction will take more of an actual time in the end, because compacting should not be a bottleneck anymore, independent of how much CPU time it takes.

Reply:
>compacting should not be a bottleneck anymore, independent of how much CPU time it takes

Why? If we limit compactors here and there, there is uncertainty how much slower it will be 

We need to find out where compaction limits are in this set up.

Reply:
> Why? If we limit compactors here and there, there is uncertainty how much slower it will be

* We don't write samples and headers to LSM, but value log
* Amount of data we write is far less than with inverted index